cms/cms_toolbars.py

```diff
@@ -313,13 +313,16 @@ def add_logout_button(self, parent):
     def add_language_menu(self):
         if settings.USE_I18N and not self._language_menu:
             self._language_menu = self.toolbar.get_or_create_menu(LANGUAGE_MENU_IDENTIFIER, _('Language'), position=-1)
-            language_changer = getattr(self.request, '_language_changer', DefaultLanguageChanger(self.request))
+            page = self.request.current_page
```

Comment:
There seemed to be effort previously to not use page. I suspect that this can be used by other content types that use the toolbar, I'll need to check.

Reply:
I'm seeing _language_changer used in a lot of places, can I check that you're also testing the changes with the CMS in standalone and not with all of the additional addons installed? Most of the tests in this area are isolating the functionality so they'll still pass as they don't appear to be touching this add_language_menu method.

Reply:
If I put in an `raise Exception` here it will fail in 97 different places. So `add_language_menu` is certainly being touched in quite a few different tests. 

> There seemed to be effort previously to not use page. I suspect that this can be used by other content types that use the toolbar, I'll need to check.

I don't really follow, to not use page? as in self.request.current_page?

If self.request.current_page happens to be None, because the url wasn't able to resolve. We fallback to using DefaultLanguageChanger which will give you the live url. So it would fail gracefully.

Reply:
@aiky30 To follow up on what will happen when no page exists such as for an alias placeholder. 

In the try ... except block will get an attribute error. So we catch the attribute error and use DefaultLanguageChanger instead. 

That will be called with the following code. 

https://github.com/divio/django-cms/blob/release/4.0.x/menus/utils.py#L151-L160

So it will give the correct url for the alias placeholder.

Reply:
Looking further into this this change is made to the BasicToolbar that won't always have a page hence the solution already in place here.